### PR TITLE
[TTS] Create EnCodec training recipe

### PR DESCRIPTION
**Pull request by rlangman**
# What does this PR do ?

This is an experimental draft for the EnCodec training recipe based on https://github.com/NVIDIA/NeMo/compare/encodec?expand=1

The PR is quite large, but has been minimized to what is necessary to run the training end to end and get a good result. It is still missing some features from the original EnCodec code that may be added in later. Including:

- Data augmentation in data loader
- Multi-resolution training (might not need to support as a feature)
- Streaming /w causal convolutions (might not need to support as a feature)
- Gradient balancer (might be unnecessary if we tune model and scale losses properly).

Note that the VQ implementation does not work with FP16 (loss always becomes NaN). I have not tested the distributed training.

**Collection**: [TTS]

# Changelog 
- Add EnCodec training recipe

# Usage

Example script invocation

```bash
python /home/rlangman/Code/NeMo/examples/tts/encodec.py
--config-path=/home/rlangman/Code/NeMo/examples/tts/conf/encodec
--config-name=encodec.yaml
max_epochs=100
weighted_sampling_steps_per_epoch=1000
batch_size=32
trainer.devices=1
log_dir=/home/rlangman/exps/dns4_jamendo_24khz/EnCodec/test_vq/samples
exp_manager.exp_dir=/home/rlangman/exps/dns4_jamendo_24khz
exp_manager.create_wandb_logger=true
+exp_manager.version=test_vq
+exp_manager.wandb_logger_kwargs.project=encodec_dns4_jamendo_24khz
+exp_manager.wandb_logger_kwargs.name=test_vq
+train_ds_meta.dns4.manifest_path=/home/rlangman/data1/data/dns4/train_manifest.json
+train_ds_meta.dns4.audio_dir=/home/rlangman/data1/data/dns4/audio_24khz
+train_ds_meta.dns4.sample_weight=0.5
+val_ds_meta.dns4.manifest_path=/home/rlangman/data1/data/dns4/val_manifest.json
+val_ds_meta.dns4.audio_dir=/home/rlangman/data1/data/dns4/audio_24khz
+log_ds_meta.dns4.manifest_path=/home/rlangman/data1/data/dns4/log_manifest.json
+log_ds_meta.dns4.audio_dir=/home/rlangman/data1/data/dns4/audio_24khz
+train_ds_meta.jamendo.manifest_path=/home/rlangman/data1/data/jamendo/train_manifest.json
+train_ds_meta.jamendo.audio_dir=/home/rlangman/data1/data/jamendo/audio_24khz
+train_ds_meta.jamendo.sample_weight=0.5
+val_ds_meta.jamendo.manifest_path=/home/rlangman/data1/data/jamendo/val_manifest.json
+val_ds_meta.jamendo.audio_dir=/home/rlangman/data1/data/jamendo/audio_24khz
+log_ds_meta.jamendo.manifest_path=/home/rlangman/data1/data/jamendo/log_manifest.json
+log_ds_meta.jamendo.audio_dir=/home/rlangman/data1/data/jamendo/audio_24khz
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

### CODE REVIEW

**Comment by titu1994** [comment]
Please do not use Hydra nested configs in NeMo. NeMo strictly follows the 1 yaml config per model standard. Yes there is some duplication, but that has been handled so far cleanly by every domain with the use of clearly defined configs, comments in yaml and simple structure of code.

Multiple config is explicitly avoided because 

1) It requires user to have the entire directly synced to properly use configs - if even one file is modified there is no way to easily check or share with us to debug
2) The user has to mentally stitch together configs from multiple files in order to compose their model - this makes it harder to know for sure what is going into your config.

**Comment by rlangman** [comment]
@titu1994 

>Please do not use Hydra nested configs in NeMo. NeMo strictly follows the 1 yaml config per model standard.

I will review these concerns with the TTS group next week.

I am also slightly surprised to hear that, since other vocoder recipes (HiFi-GAN, UnivNet) were already using nested configs. https://github.com/NVIDIA/NeMo/tree/main/examples/tts/conf/univnet

>1. It requires user to have the entire directly synced to properly use configs - if even one file is modified there is no way to easily check or share with us to debug
>2. The user has to mentally stitch together configs from multiple files in order to compose their model - this makes it harder to know for sure what is going into your config.

Aren't both of these covered by the `hparams.yaml` file in the experiment directory, which contains the final aggregated config?

>Yes there is some duplication, but that has been handled so far cleanly by every domain with the use of clearly defined configs, comments in yaml and simple structure of code.

For TTS in general I would need help figuring alternatives. The current TTS standard is duplicating a 100-300 line config file and changing 2 lines. This has become unmaintainable for combinations of sampling rate, language, g2p (eg. grapheme, IPA, ARPABET, language-specific), and model architecture variants.

But for discussion lets stick to the simpler case within this PR. We have 4 sample rates to support, and will need to eventually add at least 2-4 model variants (eg. original EnCodec, and domain optimized variants for TTS, ASR, etc). How would we restructure the config and code here to cleanly support that?

Options I could think of would be:

1. Adding wrappers within the Python code itself to do what would otherwise be done in the config. For example, model class or encoder/decoder could take in sampling rate and map them to appropriate upsample/downsample dimensions.
2. Just provide 1 example config, and tell user in documentation/tutorials/comments how to update it themselves for different configurations.

**Comment by titu1994** [comment]
> Aren't both of these covered by the hparams.yaml file in the experiment directory, which contains the final aggregated config?

Yes, that's after the model is built and already training. You have to mentally stitch together that config from multiple files first before you run, and hopefully all of your overrides are correct and pick the correct subset of files. 

> But for discussion lets stick to the simpler case within this PR. We have 4 sample rates to support, and will need to eventually add at least 2-4 model variants (eg. original EnCodec, and domain optimized variants for TTS, ASR, etc). How would we restructure the config and code here to cleanly support that?

ASR follows approach (2) you described. 

You pick one sample rate as default. User overrides it from hydra or manually before building model.

ASR has 5-6 different model sizes for each model, we don't build one config for each size, we have one standard model (usually large variant) and a table at the top explanain the config changes necessary for each size. 

Users have generally not had major issues with this. 

We tend to keep one config per "archetype" of a model. For example - there are CTC RNNT and Hybrid losses x Character or subword tokenization. That's 6 configs per model - but each is independent of the other and in different folders (only 2 per folder, char or subword config). 

While there is repeated code, it is a single file that contains the entire information of what is going into the model with all overrides visible and changeable.

File: nemo/collections/tts/losses/encodec_loss.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+    @property
+    def output_types(self):
+        return {"loss": NeuralType(elements_type=LossType())}
+
+    @typecheck()
+    def forward(self, disc_scores):
+        loss = 0.0
+        for disc_score in disc_scores:
+            loss = loss + torch.mean((1 - disc_score) ** 2)
```
[the remark this answers is not in the document]
Comment (rlangman) [reply nemo/collections/tts/losses/encodec_loss.py:134]: Changed

File: nemo/collections/tts/modules/encodec_modules.py [hunk 2]
```diff
@@ -0,0 +1,450 @@
...
+        }
+
+    def remove_weight_norm(self):
+        self.pre_conv.remove_weight_norm()
+        for res_block in self.res_blocks:
+            res_block.remove_weight_norm()
+        for down_sample_conv in self.down_sample_conv_layers:
+            down_sample_conv.remove_weight_norm()
+
+    # TODO: Add masking
```
[the remark this answers is not in the document]
Comment (rlangman) [reply nemo/collections/tts/modules/encodec_modules.py:246]: Added masking.

File: nemo/collections/tts/losses/encodec_loss.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+        loss /= len(self.mel_features)
+
+        return loss
+
+
+class FeatureMatchingLoss(Loss):
+    @property
+    def input_types(self):
+        return {
+            "fmaps_real": [[NeuralType(elements_type=VoidType())]],
```
[the remark this answers is not in the document]
Comment (rlangman) [reply nemo/collections/tts/losses/encodec_loss.py:90]: I am not sure what information will be useful on the axes/dimensions. The feature matching loss runs on a nested list of the intermediate activation outputs of the discriminators, which all have different dimensions beyond the batch size.

**Review by redoctopus — DISMISSED** [review]
LGTM

File: examples/tts/conf/audio_codec/encodec.yaml [hunk 4]
```diff
@@ -0,0 +1,159 @@
+# This config contains the default values for training 24khz EnCodec model
+# If you want to train model on other dataset, you can change config values according to your dataset.
+# Most dataset-specific arguments are in the head of the config file, see below.
+
+name: AudioCodec
```
[the remark this answers is not in the document]
Comment (rlangman) [reply examples/tts/conf/audio_codec/encodec.yaml]: Changed

File: nemo/collections/tts/losses/audio_codec_loss.py [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+        super(MaskedMAELoss, self).__init__(loss_fn=loss_fn, loss_scale=loss_scale)
+
+
+class MaskedMSELoss(MaskedLoss):
+    def __init__(self, loss_scale: float = 1.0):
+        loss_fn = torch.nn.MSELoss(reduction='none')
+        super(MaskedMSELoss, self).__init__(loss_fn=loss_fn, loss_scale=loss_scale)
+
+
+class AudioLoss(Loss):
```
[the remark this answers is not in the document]
Comment (rlangman) [reply nemo/collections/tts/losses/audio_codec_loss.py]: Changed

File: nemo/collections/tts/models/audio_codec.py [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+            self.vector_quantizer = instantiate(cfg.vector_quantizer)
+        else:
+            self.vector_quantizer = None
+
+        self.audio_decoder = instantiate(cfg.audio_decoder)
+        self.discriminator = instantiate(cfg.discriminator)
+
+        mel_loss_dim = cfg.get("mel_loss_dim", 64)
+        mel_loss_resolutions = cfg.mel_loss_resolutions
+        self.audio_loss_scale = cfg.get("audio_loss_scale", 0.1)
```
[the remark this answers is not in the document]
Comment (rlangman) [reply nemo/collections/tts/models/audio_codec.py]: To make it more consistent, I set all of the weights to 1.0 and modified it to 0.1 in encodec.yaml.
